src/lake/Lake/Util/Store.lean

```diff
@@ -3,33 +3,46 @@ Copyright (c) 2022 Mac Malone. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Mac Malone
 -/
+import Lake.Util.Family
+
 namespace Lake
 
 /-- A monad equipped with a dependently typed key-value store for a particular key. -/
+class MonadStore1Of {κ : Type u} (k : κ) (α : semiOutParam $ Type v) (m : Type v → Type w) where
+  fetch? : m (Option α)
+  store : α → m PUnit
+
+export MonadStore1Of (store)
+
+/-- Similar to `MonadStore1Of`, but `α` is an `outParam` for convenience. -/
 class MonadStore1 {κ : Type u} (k : κ) (α : outParam $ Type v) (m : Type v → Type w) where
   fetch? : m (Option α)
   store : α → m PUnit
 
-export MonadStore1 (fetch? store)
+export MonadStore1 (fetch?)
+
+instance [MonadStore1Of k α m] : MonadStore1 k α m where
+  fetch? := MonadStore1Of.fetch? k
+  store := MonadStore1Of.store k
 
 /-- A monad equipped with a dependently typed key-object store. -/
-class MonadDStore (κ : Type u) (β : outParam $ κ → Type v) (m : Type v → Type w) where
+class MonadDStore (κ : Type u) (β : semiOutParam $ κ → Type v) (m : Type v → Type w) where
   fetch? : (key : κ) → m (Option (β key))
   store : (key : κ) → β key → m PUnit
 
-instance [MonadDStore κ β m] : MonadStore1 k (β k) m where
```

Comment:
As I mentioned in the Zulip, we don't always want to use the `Family` instance (due to it requiring a `cast`).  Why did you remove this instance (and, I imagine, add the general type family instance)?

Also, from a conceptual level, there should not be a reason why `MonadStore` (which is just about equipped a monad with a key-value store) should depend on the type family implementation. That is, replace this instance with the family is conceptually equivalent to doing the same to `MonadState`, `MonadReader`, etc.

Reply:
Mario Carneiro said that it was compiled to a no-op, so that there would be no overhead. But I'll happily to go back, replacing the general instance of `FamilyDef` with the constant instance of `FamilyDef`, and adding this `MonadStore1` instance back in.

The problem here is that this instance doesn't interact well with unification, and for that I'm "misusing" the `FamilyDef`.

Reply:
> Mario Carneiro said that it was compiled to a no-op, so that there would be no overhead. 

Yes. The concern I've heard of has to do with reasoning about things with casts. Admittedly, that is not likely relevant for Lake use of this class at the moment, but is still probably something worth avoiding if possible,

Reply:
> The problem here is that this instance doesn't interact well with unification, and for that I'm "misusing" the `FamilyDef`.

I am still confused as to what exactly is the problem here. Is it that unification has a hard time solving for  `β` and the `FamilyDef` is helping it along? That is, `FamilyDef` actually is just generally useful for solving these dependently-typed unification problems? If so, to make it more generically useful and get around the potential problem with `cast`, we could move `ofFamily` and `toFamily` into the type class itself and make them avoid a `cast` in the general/constant type family case.

Reply:
Yes, the issue arises because `β` is a higher order metavariable that can be inferred via `semiOutParam`s. Unification has a hard time solving for `β`, and in the type family instance it doesn't need to solve for `β` via unification, and instead `β` is inferred as a `semiOutParam` from `MonadDStore`.

Reply:
A different solution to this problem is to completely get rid of `MonadStore1`, and instead use `MonadDStore`. I don't actually know why we needed `MonadStore1` in the first place 😅